File: apps/server/src/chat/tools/chat-tools.ts
import { tool } from "@langchain/core/tools";
import { z } from "zod";

/**
 * LangChain 工具定义
 * AI 通过调用这些工具来"插入"结构化 UI Block
 */

/** 渲染图片工具 */
export const renderImageTool = tool(
  async (input) => {
    return JSON.stringify({
      type: "image",
      content: {
        url: input.url,
        alt: input.alt || "",
        caption: input.caption || "",
      },
    });
  },
  {
    name: "render_image",
    description:
      "在回复中插入一张图片。当你需要展示图片、图表截图、示意图或任何视觉内容时使用此工具。",
    schema: z.object({
      url: z.string().describe("图片的 URL 地址"),
      alt: z.string().optional().describe("图片的替代文本描述"),
      caption: z.string().optional().describe("图片下方的说明文字"),
    }),
  },
);

/** 渲染卡片工具 */
export const renderCardTool = tool(
  async (input) => {
    return JSON.stringify({
      type: "card",
      content: {
        title: input.title,
        description: input.description || "",
        imageUrl: input.imageUrl || "",
        actions: input.actions || [],
      },
    });
  },
  {
    name: "render_card",
    description:
      "在回复中插入一张信息卡片。当你需要展示产品信息、摘要、推荐内容或任何结构化信息时使用此工具。",
    schema: z.object({
      title: z.string().describe("卡片标题"),
      description: z.string().optional().describe("卡片描述文字"),
      imageUrl: z.string().optional().describe("卡片封面图片 URL"),
      actions: z
        .array(
          z.object({
            label: z.string().describe("按钮文字"),
            url: z.string().optional().describe("点击跳转的链接"),
            variant: z
              .enum(["primary", "secondary"])
              .optional()
              .describe("按钮样式"),
          }),
        )
        .optional()
        .describe("卡片上的操作按钮列表"),
    }),
  },
);

/** 渲染图表工具 */
export const renderChartTool = tool(
  async (input) => {
    return JSON.stringify({
      type: "chart",
      content: {
        chartType: input.chartType,
        title: input.title || "",
        data: input.data,
        xKey: input.xKey || "name",
        yKey: input.yKey || "value",
        description: input.description || "",
      },
    });
  },
  {
    name: "render_chart",
    description:
      "在回复中插入一个数据图表。当你需要可视化数据、展示趋势或比较数值时使用此工具。",
    schema: z.object({
      chartType: z.enum(["bar", "line", "pie", "area"]).describe("图表类型"),
      title: z.string().optional().describe("图表标题"),
      data: z
        .array(z.record(z.unknown()))
        .describe("图表数据数组，每项是一个包含维度和度量的对象"),
      xKey: z.string().optional().describe("X 轴对应的字段名"),
      yKey: z.string().optional().describe("Y 轴对应的字段名"),
      description: z.string().optional().describe("图表的说明文字"),
    }),
  },
);

/** 渲染代码块工具 */
export const renderCodeTool = tool(
  async (input) => {
    return JSON.stringify({
      type: "code",
      content: {
        code: input.code,
        language: input.language || "text",
        filename: input.filename || "",
      },
    });
  },
  {
    name: "render_code",
    description:
      "在回复中插入一个格式化的代码块。当你需要展示代码示例、配置文件或任何技术内容时使用此工具。",
    schema: z.object({
      code: z.string().describe("代码内容"),
      language: z
        .string()
        .optional()
        .describe("编程语言，如 javascript, python, typescript 等"),
      filename: z.string().optional().describe("文件名"),
    }),
  },
);

/** 渲染表格工具 */
export const renderTableTool = tool(
  async (input) => {
    return JSON.stringify({
      type: "table",
      content: {
        headers: input.headers,
        rows: input.rows,
        caption: input.caption || "",
      },
    });
  },
  {
    name: "render_table",
    description:
      "在回复中插入一个数据表格。当你需要展示对比数据、列表或任何行列结构化数据时使用此工具。",
    schema: z.object({
      headers: z.array(z.string()).describe("表头列名数组"),
      rows: z
        .array(z.array(z.string()))
        .describe("表格数据，每行是一个字符串数组"),
      caption: z.string().optional().describe("表格标题或说明"),
    }),
  },
);

/** 所有可用工具的集合 */
export const chatTools: Array<{
  name: string;
  invoke: (input: unknown) => Promise<string>;
}> = [
  renderImageTool,
  renderCardTool,
  renderChartTool,
  renderCodeTool,
  renderTableTool,
];

/** 工具名称到工具对象的映射 */
export const toolMap = Object.fromEntries(chatTools.map((t) => [t.name, t]));

File: apps/python-server/src/chat/tools/chat_tools.py
"""LangChain 工具定义"""

import json
from typing import Any, Optional, List, Dict

from langchain_core.tools import tool
from pydantic import BaseModel, Field


class RenderImageInput(BaseModel):
    """渲染图片工具输入"""

    url: str = Field(..., description="图片的 URL 地址")
    alt: Optional[str] = Field(None, description="图片的替代文本描述")
    caption: Optional[str] = Field(None, description="图片下方的说明文字")


class CardAction(BaseModel):
    """卡片操作按钮"""

    label: str = Field(..., description="按钮文字")
    url: Optional[str] = Field(None, description="点击跳转的链接")
    variant: Optional[str] = Field(None, description="按钮样式")


class RenderCardInput(BaseModel):
    """渲染卡片工具输入"""

    title: str = Field(..., description="卡片标题")
    description: Optional[str] = Field(None, description="卡片描述文字")
    imageUrl: Optional[str] = Field(None, description="卡片封面图片 URL")
    actions: Optional[List[CardAction]] = Field(None, description="卡片上的操作按钮列表")


class RenderChartInput(BaseModel):
    """渲染图表工具输入"""

    chartType: str = Field(..., description="图表类型: bar, line, pie, area")
    title: Optional[str] = Field(None, description="图表标题")
    data: List[Dict[str, Any]] = Field(
        ..., description="图表数据数组，每项是一个包含维度和度量的对象"
    )
    xKey: Optional[str] = Field(None, description="X 轴对应的字段名")
    yKey: Optional[str] = Field(None, description="Y 轴对应的字段名")
    description: Optional[str] = Field(None, description="图表的说明文字")


class RenderCodeInput(BaseModel):
    """渲染代码块工具输入"""

    code: str = Field(..., description="代码内容")
    language: Optional[str] = Field(None, description="编程语言，如 javascript, python 等")
    filename: Optional[str] = Field(None, description="文件名")


class RenderTableInput(BaseModel):
    """渲染表格工具输入"""

    headers: List[str] = Field(..., description="表头列名数组")
    rows: List[List[str]] = Field(..., description="表格数据，每行是一个字符串数组")
    caption: Optional[str] = Field(None, description="表格标题或说明")


@tool("render_image", args_schema=RenderImageInput)
def render_image(
    url: str, alt: Optional[str] = None, caption: Optional[str] = None
) -> str:
    """在回复中插入一张图片。当你需要展示图片、图表截图、示意图或任何视觉内容时使用此工具。"""
    return json.dumps(
        {
            "type": "image",
            "content": {"url": url, "alt": alt or "", "caption": caption or ""},
        }
    )


@tool("render_card", args_schema=RenderCardInput)
def render_card(
    title: str,
    description: Optional[str] = None,
    imageUrl: Optional[str] = None,
    actions: Optional[List[Dict[str, Any]]] = None,
) -> str:
    """在回复中插入一张信息卡片。当你需要展示产品信息、摘要、推荐内容或任何结构化信息时使用此工具。"""
    return json.dumps(
        {
            "type": "card",
            "content": {
                "title": title,
                "description": description or "",
                "imageUrl": imageUrl or "",
                "actions": actions or [],
            },
        }
    )


@tool("render_chart", args_schema=RenderChartInput)
def render_chart(
    chartType: str,
    data: List[Dict[str, Any]],
    title: Optional[str] = None,
    xKey: Optional[str] = None,
    yKey: Optional[str] = None,
    description: Optional[str] = None,
) -> str:
    """在回复中插入一个数据图表。当你需要可视化数据、展示趋势或比较数值时使用此工具。"""
    return json.dumps(
        {
            "type": "chart",
            "content": {
                "chartType": chartType,
                "title": title or "",
                "data": data,
                "xKey": xKey or "name",
                "yKey": yKey or "value",
                "description": description or "",
            },
        }
    )


@tool("render_code", args_schema=RenderCodeInput)
def render_code(
    code: str, language: Optional[str] = None, filename: Optional[str] = None
) -> str:
    """在回复中插入一个格式化的代码块。当你需要展示代码示例、配置文件或任何技术内容时使用此工具。"""
    return json.dumps(
        {
            "type": "code",
            "content": {
                "code": code,
                "language": language or "text",
                "filename": filename or "",
            },
        }
    )


@tool("render_table", args_schema=RenderTableInput)
def render_table(
    headers: List[str],
    rows: List[List[str]],
    caption: Optional[str] = None,
) -> str:
    """在回复中插入一个数据表格。当你需要展示对比数据、列表或任何行列结构化数据时使用此工具。"""
    return json.dumps(
        {
            "type": "table",
            "content": {"headers": headers, "rows": rows, "caption": caption or ""},
        }
    )


# 所有可用工具的列表
chat_tools = [render_image, render_card, render_chart, render_code, render_table]

# 工具名称到工具对象的映射
tool_map = {tool.name: tool for tool in chat_tools}
